app/main/views/send.py

```diff
@@ -846,38 +837,85 @@ def send_notification(service_id, template_id):
             )
         )
 
-    db_template = current_service.get_template_with_user_permission_or_403(
-        template_id, current_user
+    keys = []
+    values = []
+    for k, v in session["placeholders"].items():
+        keys.append(k)
+        values.append(v)
+
+    data = ",".join(keys)
+    vals = ",".join(values)
+    data = f"{data}\r\n{vals}"
+
+    filename = f"one-off:{current_user.name}:{uuid.uuid4()}.csv"
+    my_data = {"filename": filename, "template_id": template_id, "data": data}
+    upload_id = s3upload(service_id, my_data)
+    form = CsvUploadForm()
+    form.file.data = my_data
+    form.file.name = filename
+
+    check_message_output = check_messages(service_id, template_id, upload_id, 2)
+    if "You cannot send to" in check_message_output:
+        return check_messages(service_id, template_id, upload_id, 2)
+
+    job_api_client.create_job(
+        upload_id,
+        service_id,
+        scheduled_for="",
+        template_id=template_id,
+        original_file_name=filename,
+        notification_count=1,
+        valid="True",
     )
 
-    try:
-        noti = notification_api_client.send_notification(
-            service_id,
-            template_id=db_template["id"],
-            recipient=recipient,
-            personalisation=session["placeholders"],
-            sender_id=session.get("sender_id", None),
+    session.pop("recipient")
+    session.pop("placeholders")
+
+    # We have to wait for the job to run and create the notification in the database
+    time.sleep(0.1)
```

Comment:
`time.sleep()` calls catch my eye - if the insertion of the DB record is dependent on the job running and finishing, should we be leveraging Celery a bit more here instead to track the status of the job and kicking things off once it's reported as finished?

We'd have to dig into the [Celery documentation](https://docs.celeryq.dev/en/stable/index.html) a bit here as I can't remember off the top of my head how to do this.  It may also require some adjustments to how things are currently set up, but it might make this a bit cleaner and more resilient as well.

Reply:
I really don't know how to make a call from admin over to the api side, get notified that the job completed -- and then, that is is not even interesting info, because the job is going to complete. We will still need to query the db.  Presumably if the job completed we only need to query once, but with the sleep implementation 99.99% of the time we will only query once also.

I know sleep() is old-school and considered a code smell these days.  I'm happy to write another ticket to investigate a better approach if need be.  I'm a bit skeptical that an approach exists that is both more resilient and easier to understand. :)

Reply:
That's fair!  I realize the time and effort would likely balloon the effort out too, so another ticket makes sense here to come back to later on.

Ha!  You may very well be right, asynchronous job processing code is... not straightforward regardless. :-)